File: lib/src/test/kotlin/com/ibm/appconfiguration/android/lib/FeatureTest.java
/*
 * (C) Copyright IBM Corp. 2021.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package com.ibm.appconfiguration.android.lib;

import com.ibm.appconfiguration.android.lib.feature.models.Feature;

import org.json.JSONArray;
import org.json.JSONObject;
import org.junit.Test;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

public class FeatureTest {

    Feature sut;

    public void setUpStringFeature(Feature.FeatureType type, Object disabled, Object enaabled, Boolean isEnabled) {

        JSONObject feature = new JSONObject();
        try {
            feature.put("name","defaultFeature");
            feature.put("feature_id","defaultfeature");
            feature.put("type",type.toString());
            feature.put("disabled_value",disabled);
            feature.put("enabled_value",enaabled);
            feature.put("isEnabled",isEnabled);
            feature.put("segment_exists", false);
            feature.put("segment_rules",new JSONArray());

        } catch (Exception e) {
            System.out.println(e);
        }
        this.sut = new Feature(feature);
    }

    @Test
    public void testStringFeature() {
        setUpStringFeature(Feature.FeatureType.STRING, "unknown user","Org user", true );
        assertEquals(sut.getFeatureDataType(), Feature.FeatureType.STRING);
        assertEquals(sut.getFeatureName(), "defaultFeature");
        assertEquals(sut.getFeatureId(), "defaultfeature");
        assertEquals(sut.isEnabled(), true);
        assertEquals((String) sut.getCurrentValue(), "Org user");
    }

    @Test
    public void testBooleanFeature() {
        setUpStringFeature(Feature.FeatureType.BOOLEAN, false,true , true);
        assertEquals(sut.getFeatureDataType(), Feature.FeatureType.BOOLEAN);
        assertEquals(sut.getFeatureName(), "defaultFeature");
        assertEquals(sut.getFeatureId(), "defaultfeature");
        assertEquals(sut.isEnabled(), true);
        assertEquals((Boolean) sut.getCurrentValue(), true);

    }

    @Test
    public void testNumericFeature() {
        setUpStringFeature(Feature.FeatureType.NUMERIC, 20,50, false );
        assertEquals(sut.getFeatureDataType(), Feature.FeatureType.NUMERIC);
        assertEquals(sut.getFeatureName(), "defaultFeature");
        assertEquals(sut.getFeatureId(), "defaultfeature");
        assertEquals(sut.isEnabled(), false);
        assertEquals((int) sut.getCurrentValue(), 20);

    }
}

File: lib/src/test/kotlin/com/ibm/cloud/appconfiguration/android/sdk/configurations/models/FeatureTest.kt
/**
 * Copyright 2021 IBM Corp. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ibm.cloud.appconfiguration.android.sdk.configurations.models

import org.json.JSONArray
import org.json.JSONObject
import org.junit.Assert.*
import org.junit.Test

class FeatureTest {

    var sut: Feature? = null

    fun setUpFeature(
        type: ConfigurationType,
        format: String?,
        disabled: Any?,
        enabled: Any?,
        isEnabled: Boolean?
    ) {
        val feature = JSONObject()
        try {
            feature.put("name", "defaultFeature")
            feature.put("feature_id", "defaultfeature")
            feature.put("type", type.toString())
            feature.put("format", format)
            feature.put("disabled_value", disabled)
            feature.put("enabled_value", enabled)
            feature.put("enabled", isEnabled)
            feature.put("segment_exists", false)
            feature.put("segment_rules", JSONArray())
        } catch (e: Exception) {
            println(e)
        }
        sut = Feature(feature)
    }

    @Test
    fun testStringFeature() {
        setUpFeature(ConfigurationType.STRING, "TEXT", "unknown user", "Org user", true)
        assertEquals(sut!!.getFeatureDataType(), ConfigurationType.STRING)
        assertEquals(sut!!.getFeatureDataFormat(), "TEXT")
        assertEquals(sut!!.getFeatureName(), "defaultFeature")
        assertEquals(sut!!.getFeatureId(), "defaultfeature")
        assertEquals(sut!!.isEnabled(), true)
        assertEquals(sut!!.getCurrentValue("pqrt", JSONObject()) as String?, "Org user")
        assertEquals(sut!!.getFeatureDisabledValue(), "unknown user")
        assertEquals(sut!!.getFeatureEnabledValue(), "Org user")
    }

    @Test
    fun testJsonFeature() {
        val enabled = JSONObject()
        enabled.put("name", "tester")
        enabled.put("description", "testing")

        val disabled = JSONObject()
        disabled.put("name", "developer")
        disabled.put("description", "coding")

        setUpFeature(ConfigurationType.STRING, "JSON", disabled, enabled, true)
        assertEquals(sut!!.getFeatureDataType(), ConfigurationType.STRING)
        assertEquals(sut!!.getFeatureDataFormat(), "JSON")
        assertEquals(sut!!.getFeatureName(), "defaultFeature")
        assertEquals(sut!!.getFeatureId(), "defaultfeature")
        assertEquals(sut!!.isEnabled(), true)
        assertEquals(
            (sut!!.getCurrentValue("pqrt", JSONObject()) as JSONObject).get("name"),
            "tester"
        )
        assertEquals((sut!!.getFeatureDisabledValue() as JSONObject).get("name"), "developer")
        assertEquals((sut!!.getFeatureEnabledValue() as JSONObject).get("name"), "tester")
    }

    @Test
    fun testYamlFeature() {
        val enabled =
            "name: tester\ndescription: testing\n---\nname: developer\ndescription: coding"
        val disabled = "name: devops\ndescription: deploying\n"
        setUpFeature(ConfigurationType.STRING, "YAML", disabled, enabled, true)
        assertEquals(sut!!.getFeatureDataType(), ConfigurationType.STRING)
        assertEquals(sut!!.getFeatureDataFormat(), "YAML")
        assertEquals(sut!!.getFeatureName(), "defaultFeature")
        assertEquals(sut!!.getFeatureId(), "defaultfeature")
        assertEquals(sut!!.isEnabled(), true)
        assertEquals(sut!!.getCurrentValue("pqrt", JSONObject()) as String?, enabled)
        assertEquals(sut!!.getFeatureDisabledValue(), disabled)
        assertEquals(sut!!.getFeatureEnabledValue(), enabled)
    }

    @Test
    fun testBooleanFeature() {
        setUpFeature(ConfigurationType.BOOLEAN, "", false, true, true)
        assertEquals(sut!!.getFeatureDataType(), ConfigurationType.BOOLEAN)
        assertEquals(sut!!.getFeatureDataFormat(), "")
        assertEquals(sut!!.getFeatureName(), "defaultFeature")
        assertEquals(sut!!.getFeatureId(), "defaultfeature")
        assertEquals(sut!!.isEnabled(), true)
        assertEquals(sut!!.getCurrentValue("pqrt", JSONObject()) as Boolean?, true)
        assertNull(sut!!.getCurrentValue("", JSONObject()))

    }

    @Test
    fun testNumericFeature() {
        setUpFeature(ConfigurationType.NUMERIC, "", 20, 50, false)
        assertEquals(sut!!.getFeatureDataType(), ConfigurationType.NUMERIC)
        assertEquals(sut!!.getFeatureDataFormat(), "")
        assertEquals(sut!!.getFeatureName(), "defaultFeature")
        assertEquals(sut!!.getFeatureId(), "defaultfeature")
        assertEquals(sut!!.isEnabled(), false)
        assertEquals(sut!!.getCurrentValue("pqrt", JSONObject()), 20)
    }

    @Test
    fun testEmptyJson() {
        sut = Feature(JSONObject())
        assert(sut!!.getFeatureName() == "")
    }
}